### PR TITLE
Social validation

### PR DESCRIPTION
**Pull request by FredPeal**
Version: The version of the package where the fix has been made.

Description:

Add an overview of the changes that your pr introduces, be short,concise and if there are any new libraries, please list them.

Features: why this feature is needed. 

Bug: the root cause and the impact it had. 

Refactor: what things are expected to improve from the refactor.

### CODE REVIEW

**Review by rwhite27 — APPROVED** [review]

File: src/Models/Sources.php [hunk 1]
```diff
@@ -58,6 +59,40 @@ public function validateAppleUser(string $identityToken) : object
...
 
     /**
+     * validation.
+     *
+     * @param  string $email
+     * @param  string $token
+     *
+     * @return bool
+     */
+    public function validation(string $email, string $token) : bool
```
Comment (kaioken) [inline src/Models/Sources.php]: @FredPeal take this out of the module and move it to a validation or service

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Providers/SocialProvider.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+namespace Canvas\Providers;
+
+use function Baka\envValue;
+use Facebook\Facebook;
+use Google_Client;
+use Phalcon\Di\DiInterface;
+use Phalcon\Di\ServiceProviderInterface;
+use Redis;
+
+class SocialProvider implements ServiceProviderInterface
```
Comment (kaioken) [inline src/Providers/SocialProvider.php:12]: @FredPeal need to add this provider to the test di

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Validations/Sources.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        $di = DI::getDefault();
+        switch ($title) {
+                case 'google':
+                        $client = $di->get('google');
+                        $payload = $client->verifyIdToken($token);
+                        if ($payload) {
+                            $userid = $payload['sub'];
+                            return $payload['email'] === $email;
+                        } else {
+                            throw new Exception('Invalid user on google validation, payload or email incorrect');
```
Comment (kaioken) [inline src/Validations/Sources.php]: @FredPeal create a specific exeception and its not define

File: src/Api/Controllers/AuthController.php [hunk 4]
```diff
@@ -346,6 +346,8 @@ public function loginBySocial() : Response
             $appleUserInfo = $source->validateAppleUser($request['social_id']);
             $request['social_id'] = $appleUserInfo->sub;
             $request['email'] = $appleUserInfo->email;
+        } elseif($source->use_validation) {
```
Comment (kaioken) [inline src/Api/Controllers/AuthController.php]: @FredPeal encapsulate this on a method

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Providers/SocialProvider.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /**
+     * @param DiInterface $container
+     */
+    public function register(DiInterface $container) : void
+    {
+        $container->setShared(
+            'facebook',
+            function (bool $prefix = true) use ($app) {
```
Comment (kaioken) [inline src/Providers/SocialProvider.php]: @FredPeal what do you use prefix for?

File: src/Validations/Sources.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                        }
+                    break;
+                case 'facebook':
+                        $fb = $di->get('facebook');
+                        $response = $fb->get('/me', $token);
+                        $user = $response->getGraphUser();
+                        if ($user) {
```
Comment (kaioken) [inline src/Validations/Sources.php]: @FredPeal your are not validating the email here ?

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Models/Sources.php [hunk 7]
```diff
@@ -70,4 +73,14 @@ public static function getByTitle(string $title) : Sources
 
         return $sourceData;
     }
+    
+    /**
+     * hasValidation
+     *
+     * @return bool
+     */
+    public function hasValidation(): bool
```
Comment (kaioken) [inline src/Models/Sources.php]: @FredPeal woulnt it be use validatoin?

**Review by kaioken — CHANGES_REQUESTED** [review]

**Review by rwhite27 — APPROVED** [review]

File: src/AuthSocial/Facebook.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Canvas\AuthSocial;
+
+use Canvas\Contracts\Auth\SocialInterface;
+use Phalcon\Di;
+use Exception;
+
+class Google implements SocialInterface
```
Comment (kaioken) [inline src/AuthSocial/Facebook.php]: @FredPeal how does it work?

File: src/Contracts/Auth/SocialInterface.php [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Canvas\Contracts\Auth;
+
+
+interface SocialInterface {
```
Comment (kaioken) [inline src/Contracts/Auth/SocialInterface.php]: @FredPeal social provider interface

File: src/AuthSocial/Facebook.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Canvas\AuthSocial;
```
Comment (kaioken) [inline src/AuthSocial/Facebook.php]: @FredPeal Auth\Social\Providers

**Review by kaioken — CHANGES_REQUESTED** [review]

**Pull request closed without merging**